### ISSUE TITLE
[BUG] gather drops the values of non-tensor stacks (the field becomes an empty TensorDict)

### ISSUE THREAD

**Issue by yupengtang** [pytorch/tensordict#1793]
## Describe the bug

`gather` drops the values of stacked non-tensor data. A `NonTensorStack` field comes back from `td.gather(dim, index)` or `torch.gather(td, dim, index)` as an empty `TensorDict`, so the strings are gone while the tensor leaves are gathered correctly. This happens on lazy stacks and on dense TensorDicts, and also for multi-dimensional gathers: a 2-D lazy stack gathered along `dim=1` with a 2-D index loses the field the same way.

Selecting the same entries with advanced indexing keeps the values, and a uniform `NonTensorData` field survives `gather`, so only per-element non-tensor stacks are affected.

## To Reproduce

```python
import torch
from tensordict import NonTensorData, NonTensorStack, TensorDict, lazy_stack

index = torch.tensor([2, 0, 1])

td = lazy_stack(
    [TensorDict(query=NonTensorData(v), x=torch.tensor(float(i))) for i, v in enumerate("abc")]
)
indexed = td[index].get("query")
gathered = td.gather(0, index)
print(type(indexed).__name__, indexed.tolist())
print(type(gathered.get("query")).__name__, list(gathered.get("query").keys()))
print(gathered.get("x").tolist())

dense = TensorDict(
    query=NonTensorStack(*[NonTensorData(v) for v in "abc"]),
    x=torch.arange(3.0),
    batch_size=[3],
)
dense_gathered = dense.gather(0, index)
print(type(dense_gathered.get("query")).__name__, list(dense_gathered.get("query").keys()))
```

```bash
NonTensorStack ['c', 'a', 'b']
TensorDict []
[2.0, 0.0, 1.0]
TensorDict []
```

The first line is `td[index]`, which keeps the strings. For a 1-D batch gathered along dim 0 with a 1-D index, `gather` selects the same entries, but the field comes back as an empty `TensorDict`, for the lazy stack and for the dense TensorDict alike.

## Expected behavior

`td.gather(0, index).get("query")` returns the same `NonTensorStack` values as `td[index]`, `['c', 'a', 'b']`, and multi-dimensional gathers select non-tensor entries by the same rule as the tensor leaves.

## System info

* tensordict installed from source at `43f9846b6fb100f49fdcce9ea7eaee6905d70db3`, built with `python setup.py build_ext --inplace` (`tensordict.__version__` reports `0.14.2`)
* Python 3.13.15, Linux
* PyTorch `2.11.0`, NumPy `2.5.2`

## Reason and Possible fixes

`torch.gather` is handled by `_gather` in `tensordict/_torch_func.py`. It rebuilds the result from the input's items and hands each value (unbatched pass-through values aside) to `_gather_tensor`, which calls `torch.gather` on it:

```python
return type(input)._new_unsafe(
    {
        key: _process_gather_value(value)
        for key, value in input.items(is_leaf=_is_leaf_nontensor)
    },
    batch_size=index.shape,
    names=names,
    device=device,
)
```

A uniform `NonTensorData` survives this because `NonTensorDataBase.__torch_function__` runs the TensorDict implementation and then restores the tensorclass with `_from_tensordict_with_copy`. `NonTensorStack` subclasses `LazyStackedTensorDict` directly and defines no `__torch_function__` of its own, so `torch.gather` on it ends up in the generic `_gather` again. A `NonTensorStack` has no keys (its values live in the stacked `NonTensorData` entries), so the rebuilt result is an empty `TensorDict`, which then replaces the field. The `out=` branch hands the same values to `_gather_tensor`.

This is the same family as #1746 (`contiguous`) and #1786 (`torch.cat`), where a lazy-stack path rebuilt the result without the non-tensor entries; #1778 fixed a related advanced-indexing problem on the same stacks. The existing `test_gather` in `test/tensorclass/test_tensorclass.py` only checks uniform string fields (`z == "foo"`), which is probably why this was not caught.

A fix could give non-tensor stacks their own path in `_gather`: select the entries value-wise following gather's index rule and restack them with `_stack_non_tensor`, the way the concatenation fix in #1786 unbinds and restacks values. I am happy to send a PR with a regression test in `test/tensordict/test_nontensor.py`, next to the `contiguous` and `cat` tests.

## Checklist

- [x] I have checked that there is no similar issue in the repo (**required**)
- [x] I have read the [documentation](https://github.com/pytorch/tensordict/tree/main/docs/) (**required**)
- [x] I have provided a minimal working example to reproduce the bug (**required**)
